File: namastack-outbox-examples/namastack-outbox-example-sns-java/src/main/java/io/namastack/demo/SnsOutboxRoutingConfiguration.java
package io.namastack.demo;

import io.namastack.demo.customer.CustomerRegisteredEvent;
import io.namastack.outbox.sns.SnsOutboxRouting;
import io.namastack.outbox.routing.selector.OutboxPayloadSelector;
import java.util.Map;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

/**
 * Configuration for SNS outbox routing using the Java builder API.
 *
 * <p>This demonstrates how to configure routing rules in pure Java:
 * <ul>
 *   <li>Route {@code CustomerRegisteredEvent} to a specific SNS topic ARN with custom message attributes</li>
 *   <li>Use a default topic ARN for all other events</li>
 * </ul>
 */
@Configuration
public class SnsOutboxRoutingConfiguration {

    @Bean
    public SnsOutboxRouting snsOutboxRouting() {
        return SnsOutboxRouting.builder()
            // Route CustomerRegisteredEvent to the customer-registrations topic
            .route(OutboxPayloadSelector.type(CustomerRegisteredEvent.class), route -> {
                route.target("arn:aws:sns:us-east-1:000000000000:customer-registrations");
                route.key((payload, metadata) -> metadata.getKey());
                route.headers((payload, metadata) -> Map.of(
                    "CustomerMail", ((CustomerRegisteredEvent) payload).getEmail()
                ));
            })
            // Default route for all other events
            .defaults(route -> {
                route.target("arn:aws:sns:us-east-1:000000000000:default-topic");
                route.key((payload, metadata) -> metadata.getKey());
                route.headers((payload, metadata) -> Map.of(
                    "eventType", payload.getClass().getSimpleName()
                ));
            })
            .build();
    }
}

File: namastack-outbox-examples/namastack-outbox-example-sns/src/main/kotlin/io/namastack/demo/SnsOutboxRoutingConfiguration.kt
package io.namastack.demo

import io.namastack.demo.customer.CustomerRegisteredEvent
import io.namastack.outbox.sns.SnsOutboxRouting
import io.namastack.outbox.sns.snsOutboxRouting
import io.namastack.outbox.routing.selector.OutboxPayloadSelector.Companion.type
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class SnsOutboxRoutingConfiguration {
    @Bean
    fun snsRouting(): SnsOutboxRouting =
        snsOutboxRouting {
            route(type(CustomerRegisteredEvent::class.java)) {
                target("arn:aws:sns:us-east-1:000000000000:customer-registrations")
                key { _, metadata -> metadata.key }
                headers { payload, _ ->
                    mapOf(
                        "CustomerMail" to (payload as CustomerRegisteredEvent).email,
                    )
                }
            }
            defaults {
                target("arn:aws:sns:us-east-1:000000000000:default-topic")
                key { _, metadata -> metadata.key }
                headers { payload, _ -> mapOf("eventType" to payload.javaClass.simpleName) }
            }
        }
}

